Spigot-Server-Patches/0615-Optimized-villager-pathfinding.patch

```diff
@@ -0,0 +1,282 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: lukas81298 <lukas81298@gmail.com>
+Date: Sat, 12 Dec 2020 16:09:32 +0100
+Subject: [PATCH] Optimized villager pathfinding
+
+The villager pathfinding searches for structures in villages a villager might navigate to (e.g. lecterns). By default, this scans all chunk sections in a horizontal range of 48 blocks. The vanilla implementation uses a very complex chain of streams performing lots of redundant computations. Furthermore, way more chunk sections are checked than actually necessary. This patch contains a rewritten version of this algorithm, which should be about 10 times faster and reduces the pressure on the garbage collector.
+Using the default configuration, the behavior should not differ from vanilla, however lowering "villageSearchRadius" in the "paper.yml" should boost the performance significantly with just a little impact on the gameplay.
+
+Furthermore, this patch adds the option to modify the "cooldown" between consecutive pathfinding calls to the same villager by modifying "villagerBehaviorMinCooldown" and
+"villagerBehaviorMaxCooldown". The cooldown will be a random number of ticks between these two values. The default config values do not change the vanilla behavior at all.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+index ae5ed3bd0b663092a4658b24cbd69d37b4e141cb..4aba6fec05851e292ef312d5f6e708b8e59c9855 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -709,4 +709,14 @@ public class PaperWorldConfig {
+     private void fixCuringExploit() {
+         fixCuringZombieVillagerDiscountExploit = getBoolean("game-mechanics.fix-curing-zombie-villager-discount-exploit", fixCuringZombieVillagerDiscountExploit);
+     }
++
++    public int villageSearchRadius = 48;
++    public int villagerBehaviorMinCooldown = 0;
++    public int villagerBehaviorMaxCooldown = 20;
++    private void villagerBehaviorSettings() {
++        villageSearchRadius = getInt("villager-behavior.village-search-radius", 48);
++        villagerBehaviorMinCooldown = getInt("villager-behavior.cooldown-min", 0);
++        villagerBehaviorMaxCooldown = getInt("villager-behavior.cooldown-max", 20);
```

Comment:
Just so it only has to be changed in one place in the future. 
```suggestion
+        villageSearchRadius = getInt("villager-behavior.village-search-radius", villageSearchRadius);
+        villagerBehaviorMinCooldown = getInt("villager-behavior.cooldown-min", villagerBehaviorMinCooldown);
+        villagerBehaviorMaxCooldown = getInt("villager-behavior.cooldown-max", villagerBehaviorMaxCooldown);
```

Reply:
Thank for this note, I've changed it